packages/zealot/src/client/client.ts

```diff
@@ -0,0 +1,229 @@
+import {PoolConfig, PoolStats, zjson} from ".."
+import {decode} from "../encoder"
+import {ResultStream} from "../query/result-stream"
+import {parseContentType} from "../util/content-type"
+import {createError} from "../util/error"
+import {
+  ClientOpts,
+  CreatePoolOpts,
+  CreatePoolResp,
+  CrossFetch,
+  LoadOpts,
+  Pool,
+  QueryOpts,
+  ResponseFormat
+} from "./types"
+
+export class Client {
+  public fetch: CrossFetch
+  public auth: string | null
+
+  constructor(public baseURL: string, opts: Partial<ClientOpts> = {}) {
+    const defaults: ClientOpts = {env: getEnv(), auth: null}
+    const options: ClientOpts = {...defaults, ...opts}
+    this.auth = options.auth || null
+    this.fetch = options.env === "node" ? nodeFetch : window.fetch.bind(window)
+  }
+
+  async version() {
+    const resp = await this.fetch(this.baseURL + "/version")
+    const content = await parseContentType(resp)
+    return resp.ok ? content : Promise.reject(createError(content))
+  }
+
+  async authMethod() {
+    // TODO
+    return Promise.resolve({} as any)
+  }
+
+  async query(query: string, opts: Partial<QueryOpts> = {}) {
+    const defaults: QueryOpts = {format: "zjson", controlMessages: true}
+    const options: QueryOpts = {...defaults, ...opts}
+    // Wrap the supplied abort with our own
+    const abort = new AbortController()
+    options.signal?.addEventListener("abort", () => {
+      abort.abort()
+    })
+
+    const resp = await this.fetch(this.baseURL + "/query", {
+      method: "POST",
+      body: JSON.stringify({query}),
+      // consolodate the creating of these fetch arguments
+      headers: {
+        Accept: getAcceptValue(options.format),
+        "Content-Type": "application/json",
+        Authorization: `Bearer ${this.auth}`
+      },
+      signal: abort.signal
+    })
+
+    if (!resp.ok) {
+      const content = await parseContentType(resp)
+      return Promise.reject(createError(content))
+    } else {
+      return new ResultStream(resp, abort)
+    }
+  }
+
+  curl(query: string, opts: Partial<QueryOpts> = {}) {
+    // add auth to this
+    const defaults: QueryOpts = {format: "zjson", controlMessages: true}
+    const options: QueryOpts = {...defaults, ...opts}
+    return `curl -X POST -d '${JSON.stringify({query})}' \\
+  -H "Accept: ${getAcceptValue(options.format)}" \\
+  -H "Content-Type: application/json" \\
+  ${this.baseURL}/query`
+  }
+
+  async createPool(name: string, opts: Partial<CreatePoolOpts> = {}) {
+    const defaults: CreatePoolOpts = {
+      order: "desc",
+      key: "ts"
+    }
+    const options: CreatePoolOpts = {...defaults, ...opts}
+    const resp = await this.fetch(this.baseURL + "/pool", {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+        Accept: getAcceptValue("zjson")
+      },
+      body: JSON.stringify({
+        name,
+        layout: {
+          order: options.order,
+          // @ts-ignore What's with all the array wrapping?
```

Comment:
I have the same question :)

Reply:
Haha, the backed uses arrays to designate field names. So [["ts"]] means the pool keys are a single field name "ts". If the key was a nested field, it would be `keys: [["id", "orig_h"]]`. If there were multiple keys it would be `keys: [["id"], ["ts"], ["my", "nested", "field"]]`.